man/spectrumctl.8
."
."     Title: spectrumctl
."    Author: Moritz Wilhelmy <crap@wzff.de>
."  Language: English
."      Date: 2010-02-21
." This document is the result of painful hand work. I still like writing manpages more than html :)
."
.TH spectrumctl 8  "February 21, 2010" "Version 0.1\-git" "Spectrum Manual"
.SH NAME
spectrumctl \- Utility to control one or more spectrum instances
.SH SYNOPSIS
.B spectrumctl
[options] action
.SH DESCRIPTION
The purpose of the \fBspectrumctl\fR utility is to control one or more Spectrum
instances. Spectrum only handles one legacy network per instance, so you need
more than one spectrum instance if you want support for more than one legacy
network. The tool offers standard tasks usually performed by
initscripts but also enables some runtime configuration of spectrum instances.
.sp
An interactive shell with comfortable tab-completion is also supported to help
with more complex administration tasks. See the \fBSHELL\fR for a detailed
explanation.
.SH ACTIONS
.sp
By default, spectrumctl acts on all transports defined in \fI/etc/spectrum/\fR.
You can use a different directory with the \fI--config-dir\fR option. If you
just want to act on a single file, you can use the \fI--config\fR option. Unless
you use \fI--config\fR, spectrumctl will silently ignore files where the
filename does not end with \fB.cfg\fR.
.sp
\fBstart\fR
.RS 4
Start all instances. This command honours the \fIenabled\fR variable in the configuration file.
.RE
.sp
\fBstop\fR
.RS 4
Stop all instances. If an instance is already stopped but the pid-file exists, it is removed.
.RE
.sp
\fBrestart\fR
.RS 4
Restart instances. This is essentially an alias for first calling \fBstop\fR and then \fBstart\fR, so users will be disconnected when invoking this command.
.RE
.sp
\fBreload\fR
.RS 4
Reload instances. This just causes spectrum instances to reopen their log-files, it does not change any runtime configuration. Unlike \fBrestart\fR, this command does not stop the transport, hence users will not notice anything.
.RE
.sp
\fBstats\fR
.RS 4
Print runtime statistics.
.RE
.sp
\fBupgrade-db\fR
.RS 4
Try to upgrade the database schema.
.RE
.sp
\fBmessage-all\fR [\fIpath\fR]
.RS 4
Message all users that are currently online. The contents of the message come from the file located at \fIpath\fR. If \fIpath\fR is omitted or \fB-\fR, this command will read the message from standard input.
.RE
.sp
\fBregister\fR \fIjid\fR \fIusername\fR [\fIpassword\fR] [\fIlanguage\fR] [\fIencoding\fR]
.RS 4
Register the user with JID \fIjid\fR and the legacy network account \fIusername\fR. If the users default \fIpassword\fR, \fIlanguage\fR and \fIencoding\fR are not given, this command will ask for these details interactivly. 

\fBNote:\fR It does not make sense to register a user across more than one transport, since a \fIusername\fR is typically only valid within a single legacy network. This command will return an error if this shell currently represents more then one transport.
.RE
.sp
\fBunregister\fR \fIjid\fR
.RS 4
Unregister the user with JID \fIjid\fR. Note that unlike the \fBregister\fR command, it makes sense to use this command accross many transports.
.RE
.sp
\fBset-vip-status\fR \fIjid\fR \fIstatus\fR
.RS 4
Set the VIP status of the user \fBjid\fR. The \fIstatus\fR should be "0" to disable VIP status and "1" to enable it.
.RE
.sp
\fBlist\fR
.RS 4
List all selected transports along with their pid, protocol and hostname. If invoked with \fI--quiet\fR, this action does not print a header line and only prints the status for transports that are not currently runnning.
.RE
.sp
\fBcron\fR
.RS 4
Output transports that recently have crashed and create coredumps and log-files.
.RE
.sp
\fBhelp\fR [\fIcmd\fR]
.RS 4
Give help about the command \fIcmd\fR. If \fIcmd\fR is omitted, print a list of available commands.
.RE
.SH OPTIONS
.RE
\fB\-c\fR \fIFILE\fR, \fB\-\-config\fR=\fIFILE\fR
.RS 4
Only act on transport configured in FILE (ignored for list)
.sp
.RE
\fB\-d\fR \fIDIR\fR, \fB\-\-config\-dir\fR=\fIDIR\fR
.RS 4
Act on all transports configured in DIR (default: \fI/etc/spectrum\fR)
.sp
.RE
\fB\-q\fR, \fB\-\-quiet\fR
.RS 4
Do not print any output. For the \fBlist\fR action, only print transports
currently not running and do not print a header line.
.sp
.RE
\fB\-h\fR, \fB\-\-help\fR
.RS 4
show a help message and exit
.sp
.RE
\fB\-\-version\fR
.RS 4
show program's version number and exit
.RE
.sp
.RE
Options for action "\fBlist\fR":
.sp
.RS 4
\fB\-\-status\fR=\fISTATUS\fR
.RS 4
Only show transports with given status. Valid values are "\fIrunning\fR", "\fIstopped\fR" and "\fIunknown\fR"
.sp
.RE
\fB\-\-machine-readable\fR
.RS 4
Output data as comma-seperated values.
.sp
.RE
\fB\-\-cron\fR
.RS 4
This option is designed to make the \fBlist\fR action suitable for periodic
cron-runs. See the \fBlist\fR action for more information.
.RE
.RE
.sp
Options for action "\fBstart\fR":
.sp
.RS 4
\fB\-\-su\fR=\fISU\fR
.RS 4
Start spectrum as this user (default: \fIspectrum\fR). Overrides the SPECTRUM_USER environment variable.
If neither \fI\-\-su\fR nor the environment variable \fBSPECTRUM_USER\fR is specified, the user defaults to 
\fIspectrum\fR.
.RE
.sp
\fB\-\-no-daemon\fR
.RS 4
Do not start spectrum as daemon.
.sp
.RE
\fB\-\-debug\fR
.RS 4
Start spectrum in debug mode. Currently this just sets the maximum size of the
core files to unlimited.

.RE
.RE
Options for action "\fBcron\fR":
.sp
.RS 4
\fB\-\-output\-dir\fR=\fIDIR\fR
.RS 4
Output files to directory \fIDIR\fR (default: \fI/root\fR).
.sp
.RE
\fB\-\-no\-backtraces\fR
.RS 4
Do not create backtraces. 
.RE
.SH SHELL
This tool also sports a fancy interactive shell that you can use to perform
multiple administration tasks at once. You can launch the shell with the
\fBshell\fR action. The shell features tab-completion for commands and a few
commands only available within the shell. 
.sp
The prompt of the spectrumctl shell represents the transports that commands will
act upon. If the shell acts upon all config-files defined in \fI/etc/spectrum\fR,
the prompt will show \fB<all transports>\fR and the JID of the current transport
otherwise.
.sp
In addition to the actions defined above, spectrumctl also supports the
following commands:
.sp
\fBload\fR \fIjid\fR
.RS 4
Only act upon the transport that serves \fIjid\fR. The prompt will be updated to reflect the change. If you use the special value \fBall\fR, all transports will be loaded again. Note that you cannot currently load config-files where the filename does not end with .cfg.
.RE
.sp
\fBexit\fR
.RS 4
Exit this shell
.RE
.RE
.SH ENVIRONMENT
The behaviour of spectrumctl can be influenced by the following environment variables:
.sp
\fBSPECTRUM_PATH\fR
.RS 4
Path where the spectrum binary is located. If omitted, spectrum is assumed to be in your PATH.
.RE
.sp
\fBSPECTRUM_USER\fR
.RS 4
The user with which spectrum is started. Overridden by the --su command line
option. 
If neither --su nor SPECTRUM_USER is specified, the user defaults to 
\fIspectrum\fR.
.RE
.SH AUTHORS
Copyright \(co 2009\-2010 by Spectrum engineers:
.sp
." template start
.RS 4
.ie n \{\h'-04'\(bu\h'+03'\c
.\}
.el \{.sp -1
.IP \(bu 2.3
.\}
Jan Kaluza <hanzz@soc.pidgin\&.im>
.RE
." template end, and once again template start
.RS 4
.ie n \{\h'-04'\(bu\h'+03'\c
.\}
.el \{.sp -1
.IP \(bu 2.3
.\}
Mathias Ertl <mati@fsinf\&.at>
.RE
." template end ;)
.RS 4
.ie n \{\h'-04'\(bu\h'+03'\c
.\}
.el \{.sp -1
.IP \(bu 2.3
.\}
Paul Aurich <paul@darkrain42\&.org>
.RE
." again template end
.sp
." TODO: Contributors section. Contributors should add themselves
.br
License GPLv3+: GNU GPL version 3 or later.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.sp
See http://gnu.org/licenses/gpl.html for more information.
.SH SEE ALSO
\fBspectrum\fP(1), \fBspectrum.cfg\fP(5)
.sp
For more information, see the spectrum homepage at http://spectrum.im/

.SH BUGS
Please submit bugs to our issue tracker at github: http://github.com/hanzz/spectrum/issues


ChangeLog
version 1.4.7 (2011-01-22):
	General:
	* Fixed 100% cpu usage when Spectrum can't connect Jabber server.
	* Resize and convert avatars before sending them into legacy network.
	  Needs libmagick++ installed during compilation.
	* Working nicknames/groups synchronization in both directions
	  using Remote Roster protoXEP. Need patched Ejabberd or extra Prosody
	  module. Check http://spectrum.im/projects/spectrum/wiki/Remote_roster.
	* Fixed crash when filetransfers have not been cancelled before XMPP user
	  disconnected.
	* Fixed bug when filetranfer got stuck sometimes.
	* Send Adhoc Commands always in the same order.
	* Answer to VCard IQ for transport contact.
	* Fixed bug when two SpectrumConversations were created for single
	  PurpleConversation and therefore messages have been sent to resource
	  with lower priority sometimes.
	* Send messages always to bare JID when resources have the same priority.
	* Send disco#info and disco#items features for all protocols. Fixes
	  bad capability hash in presences.

	ICQ:
	* Fixed avatars forwarding from Jabber to ICQ.
	* Set use_clientlogin=0 and server=login.icq.com as default values
	  for older libpurple.
	* Forward more informations in VCards. Thanks to Alexander Skovpen.
	* Fixed denying of authorization requests.

	XMPP:
	* Fixed nickname setting for Jabber MUC.
	* Send status code 110 while joining the room.

	Identica:
	* Store last message ID in database, so Spectrum doesn't send the same
	  messages again.

version 1.4.6 (2010-12-30):
	General:
	* Fixed compilation issues with older libpurple (XStatus for ICQ is
	  disabled when Spectrum is compiled with libpurple which doesn't
	  support it).

	spectrumctl:
	* Fixed syntax error caused by use of newer Python features.

version 1.4.5 (2010-12-28):
	General:
	* Map Extended Away status to Away if particular libpurple protocol
	  plugin does not support XA status.
	* Added new filetransfer_force_cache_storage=0 config variable to force
	  storing files from legacy network on server. This is also configurable
	  per user via Transport Settings adhoc commands.
	* Added "Reject all incoming authorizations" Transport Settings option.
	* Fixed forwarding of initial presence to XMPP user when more resource
	  are connected.
	* Automatic tests (spectrum_test) are temporarily disabled.

	ICQ:
	* Working XStatus forwarding from ICQ buddies to XMPP user.

	XMPP:
	* Fixed GTalk's new email notification when there are more pending emails.

	IRC:
	* Fixed forwarding of IRC users modes changes.

	Yahoo:
	* Fixed issues when receiving messages from MSN buddies using Yahoo
	  transport. Thanks to Paolo Encomienda. 

	GroupWise:
	* Added support, use protocol=groupwise in config file.

	spectrumctl:
	* No longer enforce any permissions on filetransfer_cache (it has to
	  be writable by the webserver)
	* Remove the --cron option for action 'list'
	* New action 'cron' is now intended for regular cron runs:
	  * It outputs transports that have crashed
	  * Saves backtraces, logs and the current spectrum_version (see
	    new --output-dir option)
	  * Prints location of those files
	  * Creation of backtraces can be surpressed with --no-backtraces.

version 1.4.4 (2010-11-25):
	General:
	* Error messages are sent in presences from transport contact.
	* SQL statement to check if user is VIP can be changed in config file
	  using [database] vip_statement config.
	* Fixed crash when there were buddies with non-unicode characters
	  in database

	ICQ:
	* Fixed connecting the network with libpurple 2.7.7.

version 1.4.3 (2010-10-28):
	General:
	* Fixed crash when there were buddies with non-unicode characters
	  in database
	* Faster Roster Item Exchange forwarding.
	* First presence from transport is sent when user is actually
	  logged in to legacy network.
	* Forward only presences from active resource to legacy network.
	* Send mime-types in VCards.
	* Fixed caps hash.

	ICQ:
	* Fixed logging issues.
	
version 1.4.2 (2010-10-08):
	General:
	* Fixed crash when user rejected susbcribe request after automatic
	  reconnection.
	* Fixed crash in DNSResolver when resolving was cancelled in some cases.
	* Fixed crash when conversations were not properly deleted after logout. 
	* Fixed chatstates forwarding from XMPP to legacy network.
	* Sending statistics to anyone can be disabled by setting
	  [features] statistics=0.
	* Fixed missing 'ver' attribute in presence caps after
	  "spectrumctl reload".
	* Added support for libevent. Epoll can be used instead of poll by setting
	  [service] eventloop=epoll. Use epoll only with standard libpurple
	  protocols, otherwise it doesn't have to work well.
	* Added new [registration] allowed_users=regexp config variable
	  to white-list registration usernames.

	pytransports2spectrum:
	* Fixed migration of utf8 usernames from py-transports [Nikolaus Polak].

	Bonjour:
	* Added Bonjour protocol support.

	XMPP:
	* Receiving VCards supported.

version 1.4.1 (2010-08-25):
	General:
	* Fixed memory corruption in SQLClass on x86_64.
	* Fixed memory leak in SpectrumDiscoHandler.
	* Fixed memory corruption when user settings are changed.
	* Fixed memory corruption when XMPP user tried to add legacy network
	  buddy twice.
	* Fixed crash when storing buddies settings to database when stored string
	  is NULL.
	* Fixed crash in Ad-Hoc commands when user submitted his settings after he
	  disconnected the transport.

	Twitter:
	* Store Twitter tokens in database. Fixes a bug when Spectrum was asking
	  for access on almost every login.

	Identica:
	* Fixed Identica support.

version 1.4.0 (2010-08-12):
	General:
	* Don't answer twice to disco#items stanzas.
	* Fixed memory corruption in SpectrumTimer.
	* Use transactions for storing buddies into database (storing buddies is
	  much more faster especially with SQLite).
	* Better subscriptions handling. Fixes the bug when accepting buddies with
	  capital letters or spaces.
	* Fixed disconnecting the rooms.

	Twitter:

spectrumctl/spectrumctl.py
#!/usr/bin/python -Wignore::DeprecationWarning
# -*- coding: utf-8 -*-
#
# spectrumctl can be used to control your spectrum-instances. Valid actions are
# start, stop, restart and reload. By default, spectrumctl acts on all instances
# defined in /etc/spectrum/
#
# Copyright (C) 2009, 2010 Mathias Ertl
#
# This file is part of Spectrumctl.
#
# Spectrumctl is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, see <http://www.gnu.org/licenses/>.

import os, sys
from spectrum import env, spectrum_group
from spectrum.doc import doc
from optparse import *

cmds = [ x.name for x in doc.cmds ]
description='''spectrumctl can be used to control your spectrum-instances.
Valid actions are %s and %s. By default, spectrumctl acts on all transports
defined in /etc/spectrum/.'''%( ', '.join( cmds[:-1] ), cmds[-1]  )

parser = OptionParser( usage='Usage: %prog [options] action', version='0.2', description=description)
parser.add_option( '-c', '--config', metavar='FILE',
	help = 'Only act on transport configured in FILE (ignored for list)' )
parser.add_option( '-d', '--config-dir', metavar='DIR', default='/etc/spectrum',
	help = 'Act on all transports configured in DIR (default: %default)' )
parser.add_option( '-q', '--quiet', action='store_true', default=False,
	help = 'Do not print any output' )

list_group = OptionGroup( parser, 'Options for action "list"' )
list_group.add_option( '--status', choices=['running', 'stopped', 'unknown'],
	help='Only show transports with given status. Valid values are "running", "stopped" and "unknown"' )
list_group.add_option( '--machine-readable', action='store_true', default=False,
	help='Output data as comma-seperated values' )
parser.add_option_group( list_group )

start_group = OptionGroup( parser, 'Options for action "start"' )
start_group.add_option( '--su', # NOTE: the default is set by 
	# spectrum.get_uid(). We need this so we can distinguish between
	# actually setting --su=spectrum and setting nothing at all.
	help = 'Start spectrum as this user (default: spectrum)' )
start_group.add_option( '--no-daemon', action='store_true', default=False,
	help = 'Do not start spectrum in daemon mode' )
start_group.add_option( '--debug', action='store_true', default=False,
	help = 'Start spectrum in debug mode.' )
parser.add_option_group( start_group )

coredump_group = OptionGroup( parser, 'Options for action "cron"' )
coredump_group.add_option( '--output-dir', metavar='DIR', default='/root/',
	help="Output coredumps to DIR (default: %default)" )
coredump_group.add_option( '--no-backtraces', action='store_true',
	default=False, help="Do not create any backtraces." )
parser.add_option_group( coredump_group )

options, args = parser.parse_args()
env.options = options
env.quiet = options.quiet

all_actions = [ x for x in dir( spectrum_group.spectrum_group ) if not x.startswith( '_' ) ]
if len( args ) == 0:
	print( "Please give an action (one of %s)" %(', '.join(all_actions) ) )
	sys.exit(1)

action = args[0].replace( '-', '_' )
params = args[1:]
if action not in all_actions:
	print( "Error: %s: Unknown action." %(action) )
	sys.exit(1)

group = spectrum_group.spectrum_group( options )
ret_val = getattr( group, action )( *params )
sys.exit(ret_val )
